pkg/controllers/sidecar/sync.go

```diff
@@ -11,32 +11,53 @@ import (
 	corev1 "k8s.io/api/core/v1"
 )
 
-func (mc *MemberReconciler) sync(ctx context.Context, memberService *corev1.Service) error {
+func (mc *MemberReconciler) sync(ctx context.Context, memberService *corev1.Service) (bool, error) {
 	// Check if member must decommission
 	if decommission, ok := memberService.Labels[naming.DecommissionLabel]; ok {
 		host, err := network.FindFirstNonLocalIP()
 		if err != nil {
-			return errors.Wrapf(err, "error during decommission")
+			return true, errors.Wrapf(err, "Can't get node ip")
 		}
 		// Check if member has already decommissioned
 		if decommission == naming.LabelValueTrue {
-			return nil
+			mc.logger.Debug(ctx, "Node is already decommissioned")
+			return false, nil
 		}
-		// Else, decommission member
-		if err := mc.scyllaClient.Decommission(ctx, host.String()); err != nil {
-			mc.logger.Error(ctx, "Error during decommission", "error", errors.WithStack(err))
+		opMode, err := mc.scyllaClient.OperationMode(ctx, host.String())
+		if err != nil {
+			return true, errors.Wrapf(err, "error on getting node operation mode")
 		}
-		// Confirm memberService has been decommissioned
-		if opMode, err := mc.scyllaClient.OperationMode(ctx, host.String()); err != nil || !opMode.IsDecommisioned() {
-			return errors.Wrapf(err, "error during decommission, operation mode: %s", opMode)
+		if opMode.IsLeaving() {
+			// If node is leaving, keep retrying while it left the cluster
+			return true, nil
+		} else if opMode.IsNormal() {
+			// Decommission node only if it is in normal mode
+			if err := mc.scyllaClient.Decommission(ctx, host.String()); err != nil {
+				return true, errors.Wrapf(err, "error during decommission: %s", errors.WithStack(err))
+			}
+			if opMode, err = mc.scyllaClient.OperationMode(ctx, host.String()); err != nil {
+				return true, errors.Wrapf(err, "error on getting node operation mode after decommission")
+			}
+			// Confirm memberService has been decommissioned
+			if !opMode.IsDecommisioned() {
+				return true, errors.Wrapf(err, "error during decommission, operation mode: %s", opMode)
+			}
+		} else if opMode.IsJoining() {
+			// If node is joining we need to wait till it is joined (IsNormal() == True) and then decommission it
+			mc.logger.Debug(ctx, "Can't decommission joining node")
+			return true, nil
+		} else if opMode.IsDecommisioned() {
+			mc.logger.Info(ctx, "Node is decommissioned. Skip decommission")
+			// Do nothing if it is already decommissioned
```

Comment:
```suggestion
```
This comment content is already in log message

Reply:
removed